third_party/yapf_format_diff/yapf_format_diff.py

```diff
@@ -0,0 +1,135 @@
+#!/usr/bin/env python
+#
+#===- yapf_format_diff.py - YAPF Diff Reformatter ----*- python -*--===#
+#
+# Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+#
+#===------------------------------------------------------------------------===#
+"""
+This script reads input from a unified diff and reformats all the changed
+lines. This is useful to reformat all the lines touched by a specific patch.
+Example usage:
+
+  git diff -U0 HEAD^ | yapf_format_diff.py
+
+It should be noted that the filename contained in the diff is used unmodified
+to determine the source file to update. Users calling this script directly
+should be careful to ensure that the path in the diff is correct relative to the
+current working directory.
+"""
+
+import argparse
+import difflib
+import io
+import re
+import subprocess
+import sys
+
+
+def main():
+  parser = argparse.ArgumentParser(
+      description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
+  parser.add_argument('-i',
+                      action='store_true',
+                      default=False,
+                      help='apply edits to files instead of displaying a diff')
+  parser.add_argument('-p',
+                      metavar='NUM',
+                      default=1,
+                      help='strip the smallest prefix containing P slashes'
+                      'set to 0 if passing `--no-prefix` to `git diff` or '
+                      'using svn')
+  parser.add_argument('-regex',
+                      metavar='PATTERN',
+                      default=None,
+                      help='custom pattern selecting file paths to reformat '
+                      '(case sensitive, overrides -iregex)')
+  parser.add_argument('-iregex',
+                      metavar='PATTERN',
+                      default=r'.*\.py',
+                      help='custom pattern selecting file paths to reformat '
+                      '(case insensitive, overridden by -regex)')
+  parser.add_argument('-sort-includes',
+                      action='store_true',
+                      default=False,
+                      help='let the formatter sort include blocks')
```

Comment:
This argument doesn't apply to yapf. More generally, what about only making these the args to the diff tool itself and just forwarding any other args. You can use argparse.parse_known_args to only parse the args it knows about and then forward the rest. https://docs.python.org/3.6/library/argparse.html#argparse.ArgumentParser.parse_known_args

It looks like this is actually trying to deal with `-i` although I'm not sure it really makes a ton of sense. We can leave that as-is though. That would also allow forwarding through `--diff` and such.

Reply:
Good idea! I added support for unknown args and support for `clang-format`.